service/service.ml

```diff
@@ -79,42 +81,50 @@ module Make (Opam_repo : Opam_repository_intf.S) = struct
       let request_str =
         Worker.Solve_request.to_yojson request |> Yojson.Safe.to_string
       in
       let request_str =
         Printf.sprintf "%d\n%s" (String.length request_str) request_str
       in
-      Lwt_io.write worker#stdin request_str >>= fun () ->
-      Lwt_io.read_line worker#stdout >>= fun time ->
-      Lwt_io.read_line worker#stdout >>= fun len ->
-      match Astring.String.to_int len with
-      | None -> Fmt.failwith "Bad frame from worker: time=%S len=%S" time len
-      | Some len -> (
-          let buf = Bytes.create len in
-          Lwt_io.read_into_exactly worker#stdout buf 0 len >|= fun () ->
-          let results = Bytes.unsafe_to_string buf in
-          match results.[0] with
-          | '+' ->
-              Log.info log "%s: found solution in %s s" id time;
-              let packages =
-                Astring.String.with_range ~first:1 results
-                |> Astring.String.cuts ~sep:" "
-              in
-              Ok packages
-          | '-' ->
-              Log.info log "%s: eliminated all possibilities in %s s" id time;
-              let msg = results |> Astring.String.with_range ~first:1 in
-              Error msg
-          | '!' ->
-              let msg = results |> Astring.String.with_range ~first:1 in
-              Fmt.failwith "BUG: solver worker failed: %s" msg
-          | _ -> Fmt.failwith "BUG: bad output: %s" results)
+      let process =
+        Lwt_io.write worker#stdin request_str >>= fun () ->
+        Lwt_io.read_line worker#stdout >>= fun time ->
+        Lwt_io.read_line worker#stdout >>= fun len ->
+        match Astring.String.to_int len with
+        | None -> Fmt.failwith "Bad frame from worker: time=%S len=%S" time len
+        | Some len -> (
+            let buf = Bytes.create len in
+            Lwt_io.read_into_exactly worker#stdout buf 0 len >|= fun () ->
+            let results = Bytes.unsafe_to_string buf in
+            match results.[0] with
+            | '+' ->
+                Log.info log "%s: found solution in %s s" id time;
+                let packages =
+                  Astring.String.with_range ~first:1 results
+                  |> Astring.String.cuts ~sep:" "
+                in
+                Ok packages
+            | '-' ->
+                Log.info log "%s: eliminated all possibilities in %s s" id time;
+                let msg = results |> Astring.String.with_range ~first:1 in
+                Error msg
+            | '!' ->
+                let msg = results |> Astring.String.with_range ~first:1 in
+                Fmt.failwith "BUG: solver worker failed: %s" msg
+            | _ -> Fmt.failwith "BUG: bad output: %s" results)
+      in
+      ( Lwt_switch.add_hook (Some switch) @@ fun () ->
+        (*We kill the process when there's no response from the worker*)
+        if Lwt.state process = Lwt.Sleep then (
+          Lwt.cancel process;
+          dispose worker)
+        else Lwt.return_unit );
```

Comment:
Won't this be an issue if the `worker` fails? Just to be sure that it won't accept any failed entries.

Reply:
if the `worker` fails, the state of `process` would be different to `Lwt.Sleep`. And the function `dispose` won't fail if the process of the worker fails anyway.

Reply:
Yes, that's why I was asking. To go further in detail, I was wondering under what condition the worker promise would fail, and we want / will drop the result or if it catches it somewhere else?

Reply:
> the worker promise

Do you mean the promise `process` in _`if Lwt.state process = Lwt.Sleep then`_ ?
```

Reply:
I the `process` promise fails, It is cached by `Solver_worker.Solve_request.solve` (`worker/solver_worker.mli`). But this is a rare case, like when the it fails to communicate to the sub-process (internal-workers).

Reply:
Yes. I wanted to be sure that if the `process` returned a failing promise, it wouldn't be caught in the `else` of the `if-then-else` block and return `unit`.

Reply:
Dealing with the failures is not the purpose here, it is just for the switch. The current function(_`let process ~switch ~log ~id request worker`_) return `process`, all the errors cause from the function _`let process ~switch ~log ~id request worker`_ are caught by those function calling it.